Add a `:then` filter run prefix

yaisog opened the pull request:
I would like to propose adding a `:then` filter run prefix, akin to `:else` and the `then` filter operator.

The implementation is straightforward: If there is input, replace it with the output of this filter run. The input to the filter run should be `all[tiddlers]`, or for subfilters the input to the subfilter, like `:else`.

This is an initial commit to get things rolling, created by minor modifications to the `:else` FRP. @saqimtiaz had some interesting ideas for additional variables that could be passed to the run about previous input, like a stringified version of it or an element count. This would make it behave differently from the `then[]` operator (not inherently a bad thing).

Closes #7385 (contains some examples)

To-dos:
- [x] Check that a single empty string as input is being handled like by `:else`
- [x] Pass the input along if the output of the `:then` filter run is an empty list in order not to trigger a subsequent `:else`
- [x] Add docs (with comparison to `then[]`)
- [x] Add back-reference to _then Operator_ doc
- [x] Add inline styles to doc-styles
- [x] Add tests
- [x] Improve docs
- [ ] Resolve issue with CSS in examples
- (deferred to a later PR) Decide implementation of additional variables (stringify / length vs. JSON)
- (currently not possible) Integrate with existing general documentation on FRPs 
- (currently not possible) Consolidate docs with #7368 after merging

kookma commented:
I have started some tests using the preview above Apr 3, 2023 5:46pm

**Test 1: Test01-Searchbox Casesensitive**

```
<$edit-text field=search size=50ch placeholder="enter your keywords"/>

<$let term={{!!search}}>
<$list filter="[<term>minlength[3]] :then[!is[system]search:title:casesensitive<term>]" template="$:/core/ui/ListItemTemplate"/>
</$let>
```

**PASSED**

---

**Test02-Search in List**

```
\define search-subfilter() [{$:/state/searchFields}match[all]] :then[search:*<searchTerm>] :else[search:title,caption<searchTerm>]

<$edit-text tiddler="$:/state/searchFields" field=search size=50ch placeholder="enter your keywords"/>
<$checkbox  tiddler="$:/state/searchFields" field=text checked="all" unchecked="" default=""> all fields</$checkbox>


<$let searchTerm={{{ [{$:/state/searchFields!!search}minlength[3]] }}} >
<$count filter="[!is[system]subfilter<search-subfilter>]" /> //matches//

<$list filter="[!is[system]subfilter<search-subfilter>]" template="$:/core/ui/ListItemTemplate"/>
</$let>
```

**PASSED**


---

###  The current logic
In the current implementation of `:then` the login is as below

1. if the condition met -> run the then part -> if then part is not empty ->  ignore the else part
2. if the condition met -> run the then part -> if then part is empty -> run the else part

SO, else part will run if condition does not meet or the then part is empty for met condition!

kookma commented:
@yaisog 
These are more tests! Some of them may be useful for documentation.

Example: Write a code to get a search filter and new tag from user and apply the tag to all tiddlers in search filter does not have the new tag!


**The old way**
```
\define add-tag-bulk()
<$let searchfilter={{!!searchfilter}} newTag={{!!newtag}} >
<$list filter="[subfilter<searchfilter>!is[missing]limit[1]]" variable=ignore><!--check if filter is empty -->
<$list filter="[<newTag>!is[blank]]" variable=null>
<$list filter="[subfilter<searchfilter>!tag<newTag>]">
<$fieldmangler>
<$action-sendmessage $message="tm-add-tag" $param=<<newTag>> />
</$fieldmangler>
</$list>
</$list>
</$list>
</$let>
\end


;Search filter
:<$edit-text field=searchfilter size=50 placeholder="enter a search filter"/>

;New tag
:<$edit-text field=newtag size=50 placeholder="new tag"/>

<$button actions=<<add-tag-bulk>> >
Add bulk tag
</$button>


;Results

<ul>
<$list filter={{!!searchfilter}}>
<li><<currentTiddler>> - {{!!tags}}</li>
</$list>
</ul>
```

**Use :then filter prefix** a nested :then structure

```
\define add-tag-bulk()
<$let searchfilter={{!!searchfilter}} newTag={{!!newtag}} >
<$list filter="[subfilter<searchfilter>!is[missing]]
               :then[<newTag>!is[blank]]
               :then[subfilter<searchfilter>!tag<newTag>]"
>
  <$fieldmangler>
    <$action-sendmessage $message="tm-add-tag" $param=<<newTag>> />
  </$fieldmangler>
</$list>
</$let>
\end


;Search filter
:<$edit-text field=searchfilter size=50 placeholder="enter a search filter"/>

;New tag
:<$edit-text field=newtag size=50 placeholder="new tag"/>

<$button actions=<<add-tag-bulk>> >
Add bulk tag
</$button>


;Results

<ul>
<$list filter="[tag[new]]">
<li><<currentTiddler>> - {{!!tags}}</li>
</$list>
</ul>
```

**Remarks**
- test were done using https://tiddlywiki5-git-fork-yaisog-then-filterrunprefix-jermolene.vercel.app/ above
- the second code uses nested `:then`

kookma commented:
@yaisog 
I think the empty `:then` works like

```
if cond then
    first part
if cond then
    second part
else
   last part
end if

```

The else part runs if all other parts are false! So, in `:then` case, when `:then` is empty, `:else` part runs!
This is fine except when you like to run ':else' part only and only if the first condition is false!


This can be explained in detail in documentation!

kookma commented:
I think this PR is good to be merged with 5.3.0. I used it now for 2 weeks and it seems to do its work! By the way I am speaking from an end user point of view.

@yaisog 
Do you want to make any change to this PR? I saw your todolist! My suggestion is to keep it in same line with current :else filter prefix operator, what do you think?

yaisog commented:
Hi @kookma, I think the "decide" to-dos are rather ones for @Jermolene than for me, as he has the final say in most things TW.

For his reference, the musings about what to do with an empty output are here: https://github.com/Jermolene/TiddlyWiki5/issues/7385#issuecomment-1493094415

@saqimtiaz' proposal for additional variables is here: https://github.com/Jermolene/TiddlyWiki5/issues/7385#issuecomment-1491928245. I think this could be useful. However, `__count` should be replaced by `length`, as e.g. used for the `:filter` run prefix. Making the result of the previous run available seems judicious, and a stringified list that can be separated with `enlist-input[]` is probably the best way to go about it. Alternatively, we could use the whole previous output list? I would propose a variable name without underscores or periods, though, as these are a bit of an eyesore. How about `currentInput`?

Also, I still have to check the empty input behavior. 😞

kookma commented:
Hi @yaisog, Thank you!

 I hope @Jermolene and @saqimtiaz could find free time to give their opinion/solution here.

Jermolene commented:
Thanks @yaisog @kookma @saqimtiaz. I like this idea. It would be helpful to go ahead and add docs and some tests.

Picking up the open questions that @yaisog referred to above:

> There is the one caveat that if the `:then` run results in an empty list, the `:else` part will also run, unless we design `:then` in a way that the filter run only replaces its input if it has a non-empty output and otherwise passes the input along, so that `:else` will not trigger. This might not be what subsequent runs expect, though. Also, one could put an `else` operator at the end of the `:then` run to guard against this... @saqimtiaz: How does your implementation handle this case?

I think it makes sense for the `:then` operator to only replace its input if it has a non-empty output, and otherwise pass the input along.

>  Note that it also provides two variables to the run `__input` as a stringified representation of the previous runs output and `__count`.

I think it makes sense to defer considering this possibility to give us more time to think about it. Some quick thoughts:

* Perhaps now we could pass the representation of the previous runs in JSON format as well as or instead of a list
* Rather than passing the count as a variable, we could add a `jsoncount` operator that could be used to count the entries in the JSON

yaisog commented:
> I think it makes sense to defer considering this possibility to give us more time to think about it.

Does that mean that we defer this PR, or do we merge now without the variables and add them in a later PR after we decided on a format?

Jermolene commented:
> > I think it makes sense to defer considering this possibility to give us more time to think about it.
> 
> Does that mean that we defer this PR, or do we merge now without the variables and add them in a later PR after we decided on a format?

I meant the latter: merge now without the variables and add them in a later PR after we decided on a format.

yaisog commented:
> I meant the latter: merge now without the variables and add them in a later PR after we decided on a format.

Peachy. Within that PR we might also look at all other prefixes and add something similar if it makes sense. Maybe there is a use for access to the whole input list inside `:map` or `:filter` or `:sort`? Off the top of my head I can't think of one. But we do pass the number of items for some of the prefixes...

saqimtiaz commented:
I would defer adding variables in favour of a more universal affordance to define variables in filter expressions, and proceed with this PR indendent of that.

We need tests that cover the different logic paths the code, such as no input, input, etc.

kookma commented:
> We need tests that cover the different logic paths the code, such as no input, input, etc.

I did some more tests, some may seems meaningless.

- [x]  no input 

```
<$list filter="[tag[Test]] :then[[HelloThere]tagging[]]">

</$list>
```

- [x] single input

```
<$let aTid="HelloThere" default="Learning">

<$list filter="
  [<aTid>is[tiddler]] 
  :then[<aTid>tagging[]]
  :else[<default>tagging[]]
">

</$list>
</$let>
```

- [ ] the input title (I assume this is by dsign)

```
<$list filter="[tag[HelloThere]] :then[<currentTiddler>addsuffix[/localvar]]">

</$list>
```
Should we have access to input title or no?

yaisog commented:
> Should we have access to input title, or no?

If the input is a title list, we're not iterating over that list within `:then`, so it probably doesn't make sense to set `<<currentTiddler>>` in a way like e.g. `:map` does.

kookma commented:
> If the input is a title list, we're not iterating over that list within `:then`, so it probably doesn't make sense to set `<<currentTiddler>>` in a way like e.g. `:map` does.

Fair enough! so, we can use `map` in such cases.

yaisog commented:
Doc tiddler has been added.

The same comments as in #7421 apply:

> The doc tiddler uses the overview table from #7368, but since that is not yet merged, I had to duplicate the table code. This must be taken out when #7368 is merged to avoid duplicate tables.
Also, for the same reason I cannot add this run prefix to the general run prefix documentation.
I used two additional (inline) styles that I might want to add to the doc-styles when I do the cleanup described above.

yaisog commented:
From my point of view, all currently actionable tasks are closed and this PR is ready to be reviewed.

kookma commented:
Than you @yaisog! Great job!

Some minor comments:

One visual difference is: the `:then` doc tiddler is quite different with other similar tiddlers (e.g. Map, Reduce, .. doc tiddlers) I would recommend using the same style.

1. The doc tiddler titles: `:then Filter Run Prefix` while other similar doc tiddler like map, reduce, ... do not use `:` in title. I propose `Then Filter Run Prefix` to use the same convention
2. Tags are `Filter Run Prefix`, and `Filter Syntax`. I recommend to use these like other similar tiddlers
3.  You have used a new tag `Named Filter Run Prefix`, may be it should be added to other Filter Run Prefix operators (this is not the job of this PR just for @Jermolene)
4. `New in 5.3.0` shall be put before the table (similar to current docs)
5. May be smaller examples can be added (it is important from pedagogical point of view to use short and simple examples first)

yaisog commented:
Hi @kookma, many of your concerns are advance implementations of #7368, where @pmario is currently overhauling the docs on filter syntax, including all the filter prefixes. This is just so that I don't have to make another PR to bring the `:then` docs up-to-date. My hope is that we can merge both PRs soon.

This applies to your points 1, 2, 3 and 4. Concerning your point 5, I wanted to use examples that would not better use another method like `then` and `else` operators, because I don't want to teach bad patterns just for the sake of it. All of the "simple" examples that I could think of fall into that category.

For instance, your example above
```
<$list filter="[<aTid>is[tiddler]] :then[<aTid>tagging[]] :else[<default>tagging[]]">
```
might be written more concisely using current syntax:
```
<$list filter="[<aTid>is[tiddler]else<default>tagging[]]">
```

If you have some nice short examples that cannot be implemented with current means, I'll be happy to add them.

saqimtiaz commented on editions/test/tiddlers/tests/test-prefixes-filter.js:
```diff
@@ -434,6 +434,14 @@ describe("'reduce' and 'intersection' filter prefix tests", function() {
+		expect(wiki.filterTiddlers("[[one]] [[two]] [[three]] :then[[four]]").join(",")).toBe("four");
+		expect(wiki.filterTiddlers("[[one]] :then[tag[nonexistent]]").join(",")).toBe("one");
+		expect(wiki.filterTiddlers(":then[[two]]").length).toBe(0);
+	});
```
Please add a test for a non-existent but non-blank tiddler title, e.g. `[[notatiddler]is[tiddler] :then[.....]`

yaisog replied:
Sure, done.

saqimtiaz commented on editions/tw5.com/tiddlers/system/doc-styles.tid:
```diff
@@ -286,3 +286,12 @@ ol.doc-github-contributors li {
+}
+
+.doc-op-comparison th .doc-operator {
+  background-color: unset;
```
we use tab characters for indentation rather than spaces.

yaisog replied:
Corrected with apologies.

saqimtiaz commented:
> Hi @kookma, many of your concerns are advance implementations of #7368, where @pmario is currently overhauling the docs on filter syntax, including all the filter prefixes. This is just so that I don't have to make another PR to bring the `:then` docs up-to-date. My hope is that we can merge both PRs soon.

I completely understand how time consuming it is to write documentation. The tricky thing here is that this PR becomes dependent on #7368 being merged first in the interests of consistency. While this PR involves core changes and would need a new TW release to make it live, the documentation changes in the other PR can be published to tiddlywiki.com at any time which increases the chances that that work might happen on a different timeline.

kookma commented:
> If you have some nice short examples that cannot be implemented with current means, I'll be happy to add them.

I think the below is a short example

```
<$edit-text field=search size=50ch placeholder="enter your keywords"/>

<$let term={{!!search}}>
<$list filter="[<term>minlength[3]] :then[!is[system]search:title:casesensitive<term>]" template="$:/core/ui/ListItemTemplate"/>
</$let>
```

yaisog commented:
> I think the below is a short example

Hi @kookma, thanks for the example. I have added it to the doc tiddler. I also changed the title and tags to match the current docs. These must be changed with another PR if this PR is merged before #7368. I'm not sure how to trigger that so we don't forget...

yaisog commented:
> Would you please update it to the last 5.3.0pre from master branch?

I can't:
![image](https://github.com/Jermolene/TiddlyWiki5/assets/1133745/ffe6b8b2-d328-4755-8bae-1f7f4c32ee7e)

kookma commented:
Is there any chance for merging this PR into TW 5.3.0?

saqimtiaz commented:
> > Would you please update it to the last 5.3.0pre from master branch?
> 
> I can't:

@yaisog It seems like there have been other edits to the doc-styles.tid tidder that clash with your changes. Clicking on the "resolve" button should let you manually merge the changes that have happened to that tiddler with your own changes to that tiddler.

Can you please also clarify what those CSS rules are for? Thank you.

saqimtiaz commented:
@yaisog would you be able to address the comments and resolve the conflicts in this PR this week?

If that isn't feasible, please let me know if I can fork your PR and create a resolved version.

yaisog commented:
Hi @saqimtiaz, sorry about not responding earlier. I was on vacation and thereafter busy sorting through the photos. I will try my best to get this sorted until the end of the week.

yaisog commented:
Hi @saqimtiaz, I resolved the conflicts in the doc-styles.

This PR introduced two additional styles that were needed for the operator comparison table in the doc tiddler _Then Filter Run Prefix_. I felt that such comparisons might pop up more often in the future or when reworking doc tiddlers, especially when comparing similar FRPs with their namesake filter operators, as is done here.

pmario commented on editions/tw5.com/tiddlers/system/doc-styles.tid:
```diff
@@ -285,6 +285,14 @@ ol.doc-github-contributors li {
+	table-layout: fixed;
+	width: 80%;
+}
+.doc-op-comparison th .doc-operator {
```
I think `.doc-op-comparison` classes should **not** be part of this PR. They are too usecase specific. 

There is an existing `tc-max-width-80` which sets the max-width for every container. It's defined near the [Table Utility Classes](https://github.com/Jermolene/TiddlyWiki5/blob/f4626aa69e39fb439cad674aea596ab7fc5b15c4/themes/tiddlywiki/vanilla/base.tid#L328) in the base styles.

The `table-layout: fixed;` **is new**. So there should be a new table specific utility class eg: `tc-table-layout-fixed` that's not a pretty name but it explains itself. 

The docs for Table Utility classes is at: https://tiddlywiki5-8tfwqvccw-jermolene.vercel.app/#Tables%20in%20WikiText%20CSS%20Utility%20Classes

pmario replied:
The new and needed classes may be assigned to the table itself. But there should be a separated PR with the new table-class. Once it is merged, the docs will be fixed automagically.

yaisog replied:
In CSS, `max-width: 80%` is not the same as `width: 80%`, so I could not use `tc-max-width-80`.
Creating more table utility classes may seem like a good idea, but I feel there will be endless discussions about the loss of utility due to an over-abundance of classes. These docs should not depend on another far-future PR to look right.
I could just define these styles inline if additional doc-styles are unwelcome. But then the next comparison table by someone else will look different (which might be the case anyway) or code will be needlessly duplicated.

pmario commented on editions/tw5.com/tiddlers/system/doc-styles.tid:
```diff
@@ -285,6 +285,14 @@ ol.doc-github-contributors li {
+}
+.doc-op-comparison th .doc-operator {
+	background-color: unset;
+	color: #666;
```
colors need to be used from palette values. Otherwise palette switching may make them invisible or hard to read.

yaisog replied:
There is no sensible and semantic palette entry for this application, and I would not want to define a new one for that purpose. The color `very-muted-foreground` is too light on a gray table header background.
I could use the `foreground` palette entry with some transparency, like so
```
color: <<colour foreground>>A0;
```
but I cannot be 100% sure that this will work for all palettes.

saqimtiaz commented:
@yaisog, two ways to see the comments:

1. Scroll up to about two weeks ago until you see this: ![image](https://github.com/Jermolene/TiddlyWiki5/assets/68092/f80b2cda-13c0-445e-b55c-2ddd1f657b96)
2. Go the ["Files changed" tab](https://github.com/Jermolene/TiddlyWiki5/pull/7392/files) at the top of this PR and scroll down through the code and look at comments.

yaisog commented:
@saqimtiaz, that's why I did, so at least I don't feel totally dumb right now.
This is what I see:
![image](https://github.com/Jermolene/TiddlyWiki5/assets/1133745/678685a1-9cbb-4bcb-a436-267113856f90)
There's no comment from you anywhere.
It also doesn't show a resolved conversation, so it's not that I closed it by accident. Also, I don't think that I saw the comment from your screenshot ever before.

I _do_ see @pmario's comments, though, and have responded to them.

saqimtiaz reviewed (changes requested):
@yaisog that is odd, perhaps have another look now, I am toggling the "request changes" status on my comments to see if they show up.

yaisog commented:
Hi @saqimtiaz, I see them now and will try to get to them until the end of the week.

yaisog commented on editions/tw5.com/tiddlers/filters/syntax/then Filter Run Prefix.tid:
```diff
@@ -0,0 +1,113 @@
+created: 20210618133745003
+from-version: 5.3.0
+modified: 20230506172920710
+rp-input: <<.olink all>> tiddler titles, unless otherwise determined by the first [[filter operator|Filter Operators]]
```
This was just copied from "Filter Expression":
![image](https://github.com/Jermolene/TiddlyWiki5/assets/1133745/66271d39-5843-4b1e-9666-1c9095873616)
I will remove it, but it should also be removed from there.

yaisog replied:
> The input is always all tiddler titles

Hi @saqimtiaz, also note that the input to `:and` is _not_ all tiddler titles.

saqimtiaz replied:
@yaisog my comment is about this filter run prefix specifically. There are several other run prefixes that have different input (map, reduce, etc)

saqimtiaz replied:
Using a selection constructor in a filter run prefix does not change what the input to that run was, it only ignore that input.

yaisog replied:
Hi @saqimtiaz, I agree, but then please also change the table in "Filter Expression" accordingly.

saqimtiaz replied:
I agree that the table should be corrected. Please feel free to open an issue for it if you have the opportunity.

yaisog commented on editions/tw5.com/tiddlers/filters/syntax/then Filter Run Prefix.tid:
```diff
@@ -0,0 +1,113 @@
+modified: 20230506172920710
+rp-input: <<.olink all>> tiddler titles, unless otherwise determined by the first [[filter operator|Filter Operators]]
+rp-output: the output of this filter run replaces the output of previous runs unless it is an empty list (see below)
+rp-purpose: replace any input to this filter run with its output
```
Hi @saqimtiaz, I think I need some help here. I find it exceedingly hard to wordsmith a one-liner that encompasses the FRP behavior in all edge cases. `rp-purpose` should in my opinion by as concise as possible, similar to the filter operators, since otherwise the utility of summary tiddlers such as "Filter operators" is greatly diminished.
If you have something in mind, I'll gladly use it.

saqimtiaz replied:
Perhaps something along these lines would be helpful:
`evaluate the filter run only when there is any input to it, replacing the input with its output.`

yaisog replied:
I'll adapt that. 👍

saqimtiaz commented:
@Jermolene recommend a look at this before freezing merges for the v5.3.0 release. I have reviewed the implementation and am satisfied. The documentation is worth reviewing, though it can also be improved with another PR if there are improvements identified.

Jermolene commented:
Thanks @yaisog @saqimtiaz for all your hard work, much appreciated
